core/chain_manager.go

```diff
@@ -139,14 +148,16 @@ func (bc *ChainManager) SetHead(head *types.Block) {
 	defer bc.mu.Unlock()
 
 	for block := bc.currentBlock; block != nil && block.Hash() != head.Hash(); block = bc.GetBlock(block.ParentHash()) {
-		bc.removeBlock(block)
+		DeleteBlock(bc.chainDb, block.Hash())
 	}
+	bc.headerCache.Purge()
```

Comment:
Typically the ```block.Hash() != head.Hash()``` will be hit after only a handful of blocks in regular re-orgs. Would be good to record how often and how deep re-orgs are on the live network before adding cache purging on every new head. Depending on live data, it could make sense to purge only if the common ancestor to new head is X blocks back.